File: daemon/apps/resnet18_pytorch_direct.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import os
import statistics
import time


RESNET18_INPUT_FLOATS = 1 * 3 * 224 * 224
FNV_OFFSET = 14695981039346656037
FNV_PRIME = 1099511628211


def positive_int_env(name: str, default: int) -> int:
    value = os.environ.get(name)
    if not value:
        return default
    try:
        parsed = int(value)
    except ValueError:
        return default
    return parsed if parsed > 0 else default


def cpu_list_to_string(cpu_ids: list[int]) -> str:
    if not cpu_ids:
        return "-"
    return ",".join(str(cpu_id) for cpu_id in cpu_ids)


def default_config_path() -> str:
    env_path = os.environ.get("EXAM_CONFIG_PATH")
    if env_path:
        return env_path

    for candidate in ("config/examl.yaml", "config/exam.yaml"):
        if os.path.exists(candidate):
            return candidate
    return ""


def parse_inline_cpu_list(value: str) -> list[int]:
    open_index = value.find("[")
    close_index = value.find("]", open_index + 1)
    if open_index < 0 or close_index < 0:
        return []

    cpu_ids = []
    for token in value[open_index + 1 : close_index].split(","):
        token = token.strip()
        if token:
            cpu_ids.append(int(token))
    return cpu_ids


def load_cpu_affinity_config() -> tuple[str, dict[str, list[int]]]:
    config_path = default_config_path()
    if not config_path:
        return "", {}

    cpu_affinity: dict[str, list[int]] = {}
    in_cpu_affinity = False
    cpu_affinity_indent = 0
    with open(config_path, "r", encoding="utf-8") as config_file:
        for raw_line in config_file:
            without_comment = raw_line.split("#", 1)[0].rstrip()
            stripped = without_comment.strip()
            if not stripped:
                continue

            indent = len(without_comment) - len(without_comment.lstrip(" "))
            if stripped == "cpu_affinity:":
                in_cpu_affinity = True
                cpu_affinity_indent = indent
                continue

            if in_cpu_affinity and indent <= cpu_affinity_indent:
                in_cpu_affinity = False
            if not in_cpu_affinity or ":" not in stripped:
                continue

            key, value = stripped.split(":", 1)
            cpu_affinity[key.strip()] = parse_inline_cpu_list(value)

    return config_path, cpu_affinity


def apply_pytorch_worker_cpu_affinity() -> tuple[str, list[int]]:
    config_path, cpu_affinity = load_cpu_affinity_config()
    cpu_ids = cpu_affinity.get("pytorch_worker") or cpu_affinity.get("pytorch") or []
    if cpu_ids and hasattr(os, "sched_setaffinity"):
        try:
            os.sched_setaffinity(0, set(cpu_ids))
        except OSError as exc:
            print(
                "DIRECT_CPU_AFFINITY_WARNING "
                f"cpus=[{cpu_list_to_string(cpu_ids)}] error={exc}"
            )
    return config_path, cpu_ids


def apply_realtime_scheduler() -> None:
    disabled = os.environ.get("EXAM_DISABLE_SCHED_FIFO", "")
    if disabled and disabled not in ("0", "false", "False", "off", "OFF"):
        priority = positive_int_env("EXAM_DIRECT_SCHED_FIFO_PRIORITY", 0)
        if priority > 0:
            print(
                "DIRECT_SCHED_FIFO "
                f"priority={priority} result=skipped_by_EXAM_DISABLE_SCHED_FIFO"
            )
        return

    priority = positive_int_env("EXAM_DIRECT_SCHED_FIFO_PRIORITY", 0)
    if priority <= 0:
        return

    if not hasattr(os, "sched_setscheduler") or not hasattr(os, "SCHED_FIFO"):
        print(
            "DIRECT_SCHED_FIFO_WARNING "
            f"priority={priority} error=unsupported_platform"
        )
        return

    try:
        os.sched_setscheduler(0, os.SCHED_FIFO, os.sched_param(priority))
        print(f"DIRECT_SCHED_FIFO priority={priority} result=applied")
    except OSError as exc:
        print(
            "DIRECT_SCHED_FIFO_WARNING "
            f"priority={priority} error={exc}"
        )


def configure_pytorch_runtime() -> None:
    torch.set_num_threads(positive_int_env("PYTORCH_NUM_THREADS", os.cpu_count() or 1))
    torch.set_num_interop_threads(positive_int_env("PYTORCH_INTEROP_THREADS", 1))

    torch._C._set_graph_executor_optimize(True)
    torch._C._jit_set_profiling_executor(False)
    torch._C._jit_set_profiling_mode(False)
    torch._C._jit_set_num_profiled_runs(1)


def fnv1a64(data: bytes) -> str:
    value = FNV_OFFSET
    for byte in data:
        value ^= byte
        value = (value * FNV_PRIME) & 0xFFFFFFFFFFFFFFFF
    return f"{value:016x}"


def make_input(input_id: int) -> torch.Tensor:
    state = (0x9E3779B9 ^ ((input_id * 0x85EBCA6B) & 0xFFFFFFFF)) & 0xFFFFFFFF
    values = np.empty(RESNET18_INPUT_FLOATS, dtype=np.float32)
    for index in range(RESNET18_INPUT_FLOATS):
        state = (state * 1664525 + 1013904223) & 0xFFFFFFFF
        values[index] = (
            np.float32((state >> 8) & 0xFFFF)
            / np.float32(65535.0)
            * np.float32(2.0)
            - np.float32(1.0)
        )
    return torch.from_numpy(values.reshape(1, 3, 224, 224))


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("input_id", type=int, nargs="?", default=300)
    parser.add_argument("ts_dir", nargs="?", default="artifacts/resnet18/pytorch_ts")
    parser.add_argument("warmup", type=int, nargs="?", default=5)
    parser.add_argument("iterations", type=int, nargs="?", default=10)
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    config_path, direct_cpu_ids = apply_pytorch_worker_cpu_affinity()
    apply_realtime_scheduler()

    global np
    global torch
    import numpy as np
    import torch

    print(
        "DIRECT_CPU_AFFINITY "
        f"config={config_path or '-'} "
        f"using=[{cpu_list_to_string(direct_cpu_ids)}]"
    )
    configure_pytorch_runtime()

    total = args.warmup + args.iterations
    load_start = time.perf_counter_ns()
    modules = [
        torch.jit.load(os.path.join(args.ts_dir, "sg1.pt"), map_location="cpu").eval(),
        torch.jit.load(os.path.join(args.ts_dir, "sg2.pt"), map_location="cpu").eval(),
        torch.jit.load(os.path.join(args.ts_dir, "sg3.pt"), map_location="cpu").eval(),
    ]
    load_us = (time.perf_counter_ns() - load_start) // 1000

    latencies = []
    sg_latencies = [[] for _ in modules]
    hashes = []
    with torch.inference_mode():
        for iteration in range(total):
            current_input_id = args.input_id + iteration
            tensor = make_input(current_input_id)
            start = time.perf_counter_ns()
            current_sg_latencies = []
            for module in modules:
                sg_start = time.perf_counter_ns()
                tensor = module(tensor)
                current_sg_latencies.append((time.perf_counter_ns() - sg_start) // 1000)
            latency_us = (time.perf_counter_ns() - start) // 1000
            output_hash = fnv1a64(tensor.detach().contiguous().cpu().numpy().tobytes())

            latencies.append(latency_us)
            for index, sg_latency_us in enumerate(current_sg_latencies):
                sg_latencies[index].append(sg_latency_us)
            hashes.append(output_hash)

            print(
                "DIRECT_RESULT "
                f"input_id={current_input_id} "
                f"iteration={iteration} "
                f"output_hash={output_hash} "
                f"latency_us={latency_us} "
                f"total_us={latency_us} "
                f"sg1_us={current_sg_latencies[0]} "
                f"sg2_us={current_sg_latencies[1]} "
                f"sg3_us={current_sg_latencies[2]}"
            )

    measured = latencies[args.warmup :]
    measured_hashes = hashes[args.warmup :]
    print(f"DIRECT_LOAD_US={load_us}")
    print(
        "DIRECT_THREADS "
        f"intra={torch.get_num_threads()} "
        f"interop={torch.get_num_interop_threads()}"
    )
    print("DIRECT_JIT graph_executor_optimize=on executor_mode=off profiling_mode=off num_profiled_runs=1")
    print("DIRECT_LATENCIES_US=" + " ".join(str(value) for value in measured))
    for index, values in enumerate(sg_latencies, start=1):
        measured_sg = values[args.warmup :]
        print(f"DIRECT_SG{index}_US=" + " ".join(str(value) for value in measured_sg))
    print("DIRECT_OUTPUT_HASHES=" + " ".join(measured_hashes))
    print(
        "DIRECT_STATS "
        f"count={len(measured)} "
        f"min_us={min(measured)} "
        f"avg_us={int(statistics.mean(measured))} "
        f"p50_us={int(statistics.median(measured))} "
        f"max_us={max(measured)}"
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: daemon/apps/resnet18_pytorch_direct.cpp
#include "config/cpu_affinity_config.hpp"
#include "thread_config.hpp"

#include <ATen/Parallel.h>
#include <c10/core/InferenceMode.h>
#include <torch/csrc/jit/python/update_graph_executor_opt.h>
#include <torch/csrc/jit/runtime/graph_executor.h>
#include <torch/script.h>

#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <exception>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using Clock = std::chrono::steady_clock;

constexpr const char* DEFAULT_TS_DIR = "artifacts/resnet18/pytorch";
constexpr std::size_t RESNET18_INPUT_FLOATS = 1 * 3 * 224 * 224;
constexpr std::uint64_t FNV_OFFSET = 14695981039346656037ULL;
constexpr std::uint64_t FNV_PRIME = 1099511628211ULL;

std::int64_t elapsed_microseconds(Clock::time_point start_time) {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        Clock::now() - start_time).count();
}

int positive_int_env(const char* primary, const char* fallback = nullptr) {
    const char* value = std::getenv(primary);
    if ((value == nullptr || value[0] == '\0') && fallback != nullptr) {
        value = std::getenv(fallback);
    }
    if (value == nullptr || value[0] == '\0') {
        return 0;
    }

    char* end = nullptr;
    const long parsed = std::strtol(value, &end, 10);
    if (end == value || parsed <= 0) {
        return 0;
    }
    return static_cast<int>(parsed);
}

bool bool_env_or_default(const char* name, bool default_value) {
    const char* value = std::getenv(name);
    if (value == nullptr || value[0] == '\0') {
        return default_value;
    }

    return std::string(value) != "0"
        && std::string(value) != "false"
        && std::string(value) != "False"
        && std::string(value) != "off"
        && std::string(value) != "OFF";
}

int parse_positive_int(int argc, char** argv, int index, int default_value) {
    if (argc <= index) {
        return default_value;
    }

    char* end = nullptr;
    const long parsed = std::strtol(argv[index], &end, 10);
    if (end == argv[index] || parsed <= 0) {
        return default_value;
    }
    return static_cast<int>(parsed);
}

int parse_nonnegative_int(int argc, char** argv, int index, int default_value) {
    if (argc <= index) {
        return default_value;
    }

    char* end = nullptr;
    const long parsed = std::strtol(argv[index], &end, 10);
    if (end == argv[index] || parsed < 0) {
        return default_value;
    }
    return static_cast<int>(parsed);
}

std::string optional_arg(int argc, char** argv, int index, const char* default_value) {
    if (argc <= index || argv[index][0] == '\0') {
        return default_value;
    }
    return argv[index];
}

void configure_pytorch_runtime() {
    const int num_threads =
        positive_int_env("EXAM_PYTORCH_NUM_THREADS", "PYTORCH_NUM_THREADS");
    if (num_threads > 0) {
        at::set_num_threads(num_threads);
    }

    const int interop_threads = positive_int_env(
        "EXAM_PYTORCH_INTEROP_THREADS",
        "PYTORCH_INTEROP_THREADS");
    if (interop_threads > 0) {
        at::set_num_interop_threads(interop_threads);
    }

    const bool graph_executor_optimize =
        bool_env_or_default("EXAM_TORCH_JIT_OPTIMIZE", true);
    const bool executor_mode =
        bool_env_or_default("EXAM_TORCH_JIT_EXECUTOR_MODE", false);
    const bool profiling_mode =
        bool_env_or_default("EXAM_TORCH_JIT_PROFILING_MODE", false);
    const int num_profiled_runs =
        positive_int_env("EXAM_TORCH_JIT_NUM_PROFILED_RUNS");

    torch::jit::setGraphExecutorOptimize(graph_executor_optimize);
    torch::jit::getExecutorMode().store(executor_mode);
    torch::jit::getProfilingMode().store(profiling_mode);
    if (num_profiled_runs > 0) {
        torch::jit::getNumProfiledRuns().store(
            static_cast<std::size_t>(num_profiled_runs));
    }
}

std::uint64_t fnv1a64(const void* data, std::size_t size) {
    const auto* bytes = static_cast<const unsigned char*>(data);
    std::uint64_t hash = FNV_OFFSET;
    for (std::size_t index = 0; index < size; ++index) {
        hash ^= bytes[index];
        hash *= FNV_PRIME;
    }
    return hash;
}

std::string hex64(std::uint64_t value) {
    std::ostringstream output;
    output << std::hex << std::setfill('0') << std::setw(16) << value;
    return output.str();
}

std::vector<float> make_resnet18_input(int input_id) {
    std::vector<float> input(RESNET18_INPUT_FLOATS);
    std::uint32_t state =
        0x9e3779b9U ^ (static_cast<std::uint32_t>(input_id) * 0x85ebca6bU);
    for (std::size_t index = 0; index < RESNET18_INPUT_FLOATS; ++index) {
        state = state * 1664525U + 1013904223U;
        input[index] =
            (static_cast<float>((state >> 8) & 0xFFFFU) / 65535.0F) * 2.0F
            - 1.0F;
    }
    return input;
}

std::string tensor_hash(const at::Tensor& tensor) {
    at::Tensor cpu_tensor = tensor;
    if (!cpu_tensor.device().is_cpu()) {
        cpu_tensor = cpu_tensor.to(torch::kCPU);
    }
    if (!cpu_tensor.is_contiguous()) {
        cpu_tensor = cpu_tensor.contiguous();
    }

    const std::size_t byte_size =
        static_cast<std::size_t>(cpu_tensor.numel()) * cpu_tensor.element_size();
    return hex64(fnv1a64(cpu_tensor.data_ptr(), byte_size));
}

std::string join_values(const std::vector<std::int64_t>& values) {
    std::ostringstream output;
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index != 0) {
            output << ' ';
        }
        output << values[index];
    }
    return output.str();
}

std::string join_strings(const std::vector<std::string>& values) {
    std::ostringstream output;
    for (std::size_t index = 0; index < values.size(); ++index) {
        if (index != 0) {
            output << ' ';
        }
        output << values[index];
    }
    return output.str();
}

torch::jit::script::Module load_module(const std::string& path) {
    torch::jit::script::Module module = torch::jit::load(path);
    module.to(torch::kCPU);
    module.eval();
    return module;
}

at::Tensor run_module(torch::jit::script::Module& module, const at::Tensor& input) {
    std::vector<torch::jit::IValue> inputs;
    inputs.emplace_back(input);
    torch::jit::IValue result = module.forward(inputs);
    if (!result.isTensor()) {
        throw std::runtime_error("resnet18 pytorch direct requires tensor output");
    }
    return result.toTensor();
}

} // namespace

int main(int argc, char** argv) {
    try {
        std::cout << std::unitbuf;
        std::cerr << std::unitbuf;

        const int input_id = parse_positive_int(argc, argv, 1, 300);
        const std::string ts_dir = optional_arg(argc, argv, 2, DEFAULT_TS_DIR);
        const int warmup = parse_nonnegative_int(argc, argv, 3, 5);
        const int iterations = parse_positive_int(argc, argv, 4, 10);
        const int total = warmup + iterations;

        const exam::CpuAffinityConfig cpu_affinity =
            exam::CpuAffinityConfig::load_default();
        const std::vector<int> direct_cpu_ids =
            cpu_affinity.cpus_for_pytorch_direct();
        exam::ThreadConfig::bind_to_cpus(
            "resnet18-pytorch-direct",
            direct_cpu_ids);
        std::cout << "CPP_DIRECT_CPU_AFFINITY " << cpu_affinity.summary()
                  << " using=[" << exam::cpu_ids_to_string(direct_cpu_ids)
                  << "]\n";

        const int sched_fifo_priority =
            positive_int_env("EXAM_DIRECT_SCHED_FIFO_PRIORITY");
        if (sched_fifo_priority > 0) {
            exam::ThreadConfig::set_priority(
                "resnet18-pytorch-direct",
                sched_fifo_priority);
        }

        configure_pytorch_runtime();

        const auto load_started_at = Clock::now();
        std::vector<torch::jit::script::Module> modules;
        modules.reserve(3);
        modules.push_back(load_module(ts_dir + "/sg1.pt"));
        modules.push_back(load_module(ts_dir + "/sg2.pt"));
        modules.push_back(load_module(ts_dir + "/sg3.pt"));
        const auto load_us = elapsed_microseconds(load_started_at);

        std::vector<std::int64_t> latencies;
        std::vector<std::int64_t> sg1_latencies;
        std::vector<std::int64_t> sg2_latencies;
        std::vector<std::int64_t> sg3_latencies;
        std::vector<std::string> hashes;
        latencies.reserve(total);
        sg1_latencies.reserve(total);
        sg2_latencies.reserve(total);
        sg3_latencies.reserve(total);
        hashes.reserve(total);

        c10::InferenceMode inference_mode(true);
        for (int iteration = 0; iteration < total; ++iteration) {
            const int current_input_id = input_id + iteration;
            std::vector<float> input_values = make_resnet18_input(current_input_id);
            at::Tensor tensor = torch::from_blob(
                input_values.data(),
                {1, 3, 224, 224},
                torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCPU));

            std::vector<std::int64_t> current_sg_latencies;
            current_sg_latencies.reserve(modules.size());
            const auto started_at = Clock::now();
            for (auto& module : modules) {
                const auto sg_started_at = Clock::now();
                tensor = run_module(module, tensor);
                current_sg_latencies.push_back(elapsed_microseconds(sg_started_at));
            }
            const auto latency_us = elapsed_microseconds(started_at);
            const std::string output_hash = tensor_hash(tensor);

            latencies.push_back(latency_us);
            sg1_latencies.push_back(current_sg_latencies[0]);
            sg2_latencies.push_back(current_sg_latencies[1]);
            sg3_latencies.push_back(current_sg_latencies[2]);
            hashes.push_back(output_hash);

            std::cout << "CPP_DIRECT_RESULT input_id=" << current_input_id
                      << " iteration=" << iteration
                      << " output_hash=" << output_hash
                      << " latency_us=" << latency_us
                      << " total_us=" << latency_us
                      << " sg1_us=" << current_sg_latencies[0]
                      << " sg2_us=" << current_sg_latencies[1]
                      << " sg3_us=" << current_sg_latencies[2] << '\n';
        }

        const auto measured_begin = static_cast<std::size_t>(warmup);
        std::vector<std::int64_t> measured_latencies(
            latencies.begin() + measured_begin,
            latencies.end());
        std::vector<std::int64_t> measured_sg1(
            sg1_latencies.begin() + measured_begin,
            sg1_latencies.end());
        std::vector<std::int64_t> measured_sg2(
            sg2_latencies.begin() + measured_begin,
            sg2_latencies.end());
        std::vector<std::int64_t> measured_sg3(
            sg3_latencies.begin() + measured_begin,
            sg3_latencies.end());
        std::vector<std::string> measured_hashes(
            hashes.begin() + measured_begin,
            hashes.end());

        std::cout << "CPP_DIRECT_LOAD_US=" << load_us << '\n';
        std::cout << "CPP_DIRECT_THREADS intra=" << at::get_num_threads()
                  << " interop=" << at::get_num_interop_threads() << '\n';
        std::cout << "CPP_DIRECT_JIT graph_executor_optimize="
                  << (torch::jit::getGraphExecutorOptimize() ? "on" : "off")
                  << " executor_mode="
                  << (torch::jit::getExecutorMode().load() ? "on" : "off")
                  << " profiling_mode="
                  << (torch::jit::getProfilingMode().load() ? "on" : "off")
                  << " num_profiled_runs="
                  << torch::jit::getNumProfiledRuns().load() << '\n';
        std::cout << "CPP_DIRECT_LATENCIES_US="
                  << join_values(measured_latencies) << '\n';
        std::cout << "CPP_DIRECT_SG1_US=" << join_values(measured_sg1) << '\n';
        std::cout << "CPP_DIRECT_SG2_US=" << join_values(measured_sg2) << '\n';
        std::cout << "CPP_DIRECT_SG3_US=" << join_values(measured_sg3) << '\n';
        std::cout << "CPP_DIRECT_OUTPUT_HASHES="
                  << join_strings(measured_hashes) << '\n';

        return 0;
    } catch (const std::exception& e) {
        std::cerr << "resnet18_pytorch_direct: " << e.what() << '\n';
        return 1;
    }
}
